### PR TITLE
Smart Wallet Dapp Architecture

### PR DESCRIPTION
**Pull request by dckc**
rendered: [Smart Wallet Dapp Architecture](https://dc-board-client-marshal.documentation-7tp.pages.dev/guides/getting-started/contract-rpc.html)

including diagrams, esp.

```mermaid
sequenceDiagram
  actor client

  box aqua Cosmos SDK layer
  participant chain
  end

  box darksalmon Hardened JS VM
  participant walletFactory
  participant zoe
  participant contract
  end

    client->>chain: Offer tx
    chain->>walletFactory: Offer Spec
    walletFactory->>+zoe: Offer
    walletFactory-->>chain: Offer status
    client->>+chain: vstorage query
    chain-->>-client: Offer status
    zoe->>+contract: Proposal
    contract-->>-zoe: reallocations
    zoe-->>-walletFactory: payouts
    walletFactory-->>chain: Offer status
    client->>+chain: vstorage query
    chain-->>-client: Offer status
```

based on notes from

 - [Jul 19 Office Hours: promises in terms, marshalling w/board ids, \.\.\. · Agoric/agoric\-sdk · Discussion \#8069](https://github.com/Agoric/agoric-sdk/discussions/8069)

see also initial diagrams https://github.com/Agoric/agoric-sdk/discussions/6934#discussioncomment-6686426

## Markdown lint

My editor did some markdown lint on save. Without tool support in this repo, that's perhaps not a good idea. I'm willing to back it out if it seems to noisy.

### CODE REVIEW

File: main/guides/getting-started/contract-rpc.md [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+    "denom": "uist",
+    "amount": "215000"
+  }
+]
+```
+
+They are not published redundantly in vstorage and nor does the
+smart wallet emit `balance` updates for them.
+
+To get the corresondence between certain cosmos denoms (chosen by governance)
```
Comment (WietzeSlagman) [inline main/guides/getting-started/contract-rpc.md:359]: small typo of correspondence

File: main/guides/getting-started/contract-rpc.md [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+6. Zoe ensures that the reallocations respect offer safety and then provides payouts accordingly.
+7. The client's query tells it that the payouts are available.
+
+## Signing and Broadcasting Offers
+
+One way to sign and broadcast offers is with the `agd tx ...` command. For example:
+
+```sh
+agd tx swingset wallet-action --allow-spend "$ACTION" \
+ --chain-id=agoriclocal --from=acct1
```
Comment (dckc) [inline main/guides/getting-started/contract-rpc.md:29]: note that this assumes a local chain? explain how to point it to devnet?

File: main/guides/getting-started/contract-rpc.md [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+Recall that for an agent within the JavaScript VM,
+[E(zoe).offer(...)](../../reference/zoe-api/zoe.md#e-zoe-offer-invitation-proposal-paymentkeywordrecord-offerargs) takes an `Invitation` and optionally a `Proposal` with `{ give, want }`, a `PaymentKeywordRecord`, and `offerArgs`; it returns a `UserSeat` from which we can [getPayouts()](../../reference/zoe-api/user-seat.md#e-userseat-getpayouts).
+
+![Zoe API diagram, simplified](./assets/zoe-simp.svg)
+
+In the Smart Wallet architecture, a client uses an `OfferSpec` to
+tell its `SmartWallet` how to conduct an offer.
+It includes an `invitationSpec` to say what invitation to pass to Zoe. For example:
+
+<<< @/snippets/test-marshal.js#exInvitationSpec
```
Comment (dckc) [inline main/guides/getting-started/contract-rpc.md:123]: forward reference to how we got `instance`?

explain that the code is excerpted from some snippets that provide some stuff in ways not shown?

File: main/guides/getting-started/contract-rpc.md [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+}
+```
+
+## VBank Assets and Cosmos Bank Balances
+
+Note that balances of assets such as **IST** and **BLD** are already
+available via consensus layer queries to the Cosmos SDK [bank module](https://docs.cosmos.network/main/modules/bank).
+
+```sh
+$ agd query bank balances agoric1h4d3mdvyqhy2vnw2shq4pm5duz5u8wa33jy6cl -o json | jq .balances
```
Comment (dckc) [inline main/guides/getting-started/contract-rpc.md:339]: @Wietze wrote:

> What might be a help for new developers is to provide a clear example for querying the balances of vbank assets (IST etc), now it feels a bit disjointed to have to go to the CosmosSDK to do the rest (most Agoric implementations fully abstract it away which is I think a good thing). Potentially pointing them to where this happens in wallet-UI might be enough. 

> Will there be some kind of wrapper within the agoric-sdk for this in the future potentially 

I replied:

I could see us factoring out such a wrapper... Somewhere in the ui-kit repo, perhaps.

Querying balances with just fetch is pretty straightforward... I suppose an example could be pretty small. But the page is getting long... I might want to split it up. Hm.

File: main/guides/getting-started/contract-rpc.md [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+`body:`, `slots:`, and `$1.Alleged: Gold Brand` need further explanation.
+
+### Marshalling Amounts and Instances
+
+To start with, amounts include `bigint`s. The `@endo/marshal` API handles those:
+
+<<< @/snippets/test-marshal.js#marshal-json-steroids
+
+To marshal brands and instances, recall from the discussion of [marshal in eventual send](../js-programming/eventual-send.md#e-and-marshal-a-closer-look)
+how remotables are marshalled with a translation table.
```
Comment (dckc) [inline main/guides/getting-started/contract-rpc.md:157]: add link to office hours session?

File: main/guides/getting-started/contract-rpc.md [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+translation table:
+
+<<< @/snippets/test-marshal.js#boardMarshal
+
+To reverse the process, clients can mirror the on-chain board translation
+table by synthesizing a remotable for each reference marker received:
+
+<<< @/snippets/test-marshal.js#makeBoardContext
+
+Now we can take results of vstorage queries for `Data('published.agoricNames.brand')` and `Data('published.agoricNames.instance')` unmarshal ("ingest") them:
```
Comment (dckc) [inline main/guides/getting-started/contract-rpc.md:171]: "ingest" is probably an unnecessary concept. just use `fromCapData`.

(having studied popular definitions of "serialize" and "marsha" some more, I think it's not wrong to use them with the CapData structure. I regret the renaming to `toCapData` and `fromCapData`)

File: main/guides/getting-started/contract-rpc.md [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+One way to sign and broadcast offers is with the `agd tx ...` command. For example:
+
+```sh
+agd tx swingset wallet-action --allow-spend "$ACTION" \
+ --chain-id=agoriclocal --from=acct1
+```
+
+<!-- TODO: discuss agoric wallet send? -->
+
+Another is using a wallet signing UI such as Keplr via the [Keplr API](https://docs.keplr.app/api/).
```
Comment (dckc) [inline main/guides/getting-started/contract-rpc.md:34]: A screenshot of signing with Keplr would be nice.

I looked high and low for one.
  - wallet UI docs? nope
  - docs.inter.trade? nope
  - keplr docs? nope.

File: main/guides/getting-started/contract-rpc.md [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+
+The [vstorage-viewer](https://github.com/p2p-org/p2p-agoric-vstorage-viewer) contributed by p2p is often _very_ handy:
+
+[![vstorage viewer screenshot](https://user-images.githubusercontent.com/150986/259798595-40cd22f0-fa01-43a9-b92a-4f0f4813a4f6.png)](https://p2p-org.github.io/p2p-agoric-vstorage-viewer/#https://devnet.rpc.agoric.net/|published,published.agoricNames|)
+
+:::
+
+## Specifying Offers
+
+Recall that for an agent within the JavaScript VM,
```
Comment (dckc) [inline main/guides/getting-started/contract-rpc.md:114]: "Recall" works only if Zoe has previously been introduced. Does this doc go in Getting Started or Guides?

**Comment by samsiegart** [comment]
Great stuff. In general I think all of this is crucial and important to have documented as such, and I'm just considering how to distill it. I assume the target audience is thinking "I want to make a dapp, what do I need to know?". Some of these items are lower level than most dapp developers should hopefully need to be concerned with. I think it boils down to two main categories: how to read from the chain, and how to make offers. I wonder if we could reorganize the sections as such, and at the top start with a quick and dirty code sample, maybe utilizing something in https://github.com/Agoric/ui-kit

**Comment by dckc** [comment]
Agreed, @samsiegart , it's important to focus on what devs need to know and not burden them with either too much implementation detail nor too much "big picture" stuff.

This _is_ intended to be about the overall architecture. So I'm inclined to stick with the figure to start with; I don't think a code sample works the same way. But we should also have a page that starts with a code sample. And I should probably break this page up and/or remove some detail.

File: main/guides/getting-started/contract-rpc.md (hunk 6 above)
Comment (turadg) [reply main/guides/getting-started/contract-rpc.md:177]: >"ingest" is probably an unnecessary concept. just use fromCapData.

Agree.

> (having studied popular definitions of "serialize" and "marsha" some more, I think it's not wrong to use them with the CapData structure. I regret the renaming to toCapData and fromCapData)

Disagree. 

> In computing, serialization (or serialisation) is the process of translating a [data structure](https://en.wikipedia.org/wiki/Data_structure) or [object](https://en.wikipedia.org/wiki/Object_(computer_science)) state into a format that can be stored (e.g. [files](https://en.wikipedia.org/wiki/Computer_file) in [secondary storage devices](https://en.wikipedia.org/wiki/Secondary_storage_devices), [data buffers](https://en.wikipedia.org/wiki/Data_buffer) in primary storage devices) or transmitted (e.g. [data streams](https://en.wikipedia.org/wiki/Data_stream) over [computer networks](https://en.wikipedia.org/wiki/Computer_networks)) and reconstructed later (possibly in a different computer environment).[[1]](https://en.wikipedia.org/wiki/Serialization#cite_note-1)

The `serialize` function was not doing this. It was turning one object graph (original) into another (Cap Data structure).

File: main/guides/getting-started/contract-rpc.md (hunk 7 above)
Comment (turadg) [reply main/guides/getting-started/contract-rpc.md:34]: a TODO comment  for a screenshot may be worthwhile

File: main/guides/getting-started/contract-rpc.md (hunk 8 above)
Comment (turadg) [reply main/guides/getting-started/contract-rpc.md:114]: Agree. Consider simply referencing Zoe docs.

**Review by turadg — APPROVED** [review]
This is a leap forward in the documentation. I see nothing incorrect.

It can continue to be improved more but I think this is good to merge.

File: main/guides/getting-started/contract-rpc.md (hunk 6 above)
Comment (dckc) [reply main/guides/getting-started/contract-rpc.md:177]: The Cap Data structure is straightforward to store / transmit.

**Pull request merged**